recipes-extended/xen/files/libxl-fix-flr.patch

```diff
@@ -125,66 +124,75 @@ PATCHES
      fd = open(reset, O_WRONLY);
      if (fd >= 0) {
          char *buf = GCSPRINTF(PCI_BDF, domain, bus, dev, func);
-@@ -1501,10 +1500,6 @@ skip1:
-         fclose(f);
+@@ -1596,7 +1595,8 @@ void libxl__device_pci_add(libxl__egc *e
+         goto out;
      }
- out:
+ 
+-    libxl__device_pci_reset(gc, pcidev->domain, pcidev->bus, pcidev->dev, pcidev->func);
++    /* Ugh this needs to go.  All resets must happen before any are assigned. */
++    /* libxl__device_pci_reset(gc, pcidev->domain, pcidev->bus, pcidev->dev, pcidev->func); */
```

Comment:
Based on your commit message, under what circumstances would a device be hot-added to a domain in openxt?  Currently all devices/functions are reset and assigned prior to domain start, and then reset after domain death.  I can't think of a scenario where one function gets reset independently of a 2nd function, while the guest is running.

Reply:
I might have messed this up?

For an HVM domain with a stubdom, devices are PV PCI hotplugged into the running PV stubdom, and then QMP assigned to QEMU before booting the HVM.

I thought the issue was that with "thorough reset" potentially resetting a slot, reset cannot be done while any devices are assigned.  The patch description mentioned "delayed until the domain has fully been destroy."

I thought the patch was to move all resets to only the shutdown side.  That way, any devices are reset when they move from the guest back to the system, and the potential reset issues are avoided.  Was the reset during startup supposed to remain?

If the reset during startup remains, would a thorough reset be a problem?
2 devices at startup:
Reset 1: Ok
Reset 2: Potentially bad

The devices wouldn't be in use by the HVM guest (though they would by in stubdom via pcifront), so this is probably okay?  The real guest isn't using them yet, and the stubdom shouldn't use them on its own.

Seems like the best flow would be:
1 reset all PCI devices
2 assign PCI devices
3 run - shutdown vm
4 deassign PCI
5 reset all PCI

I also wonder about cross-VM affects of the PCI thorough reset if the slot reset happens.  Is that a concern?

(libxl-fix-flr.patch gets modified a few times in subsequent patches, but not in a way relevant to this change.)

Reply:
So I was digging through this stuff today and I may have been mistaken about resets.  Devices _are_ reset before they're assigned to a domain but the libxl code that reset is attached to I believe is for like, ```xl pci-attach```, not the normal domain creation with pci devices specified in the pci list in the domain's xl.conf.

That being said, I'm not 100% sure how those pci devices in the domain xl.conf list get added to the domain during creation.  I will keep looking, because this would give us a hard answer as to whether or not devices are reset before domain start

That flow sounds correct to me.  We avoid cross-VM problems in the downstream by enforcing that all functions on the device get passed through to the same VM.  No assigning the GPU video function to one Windows guest and GPU Audio function to a different Windows guest.  That way all functions on the slot (device) are released after domain shutdown and if FLR fails, slot (device) reset should succeed.

Reply:
Domain create calls the same functions as xl pci-attach.  Domain create does:
```
libxl__add_pcis()
  loop
    libxl__device_pci_add()
```
whereas `xl pci-attach` calls:
```
libxl_device_pci_add()
  libxl__device_pci_add()
```

Upstream Xen does a reset in libxl__device_pci_add() and pci_remove_detached().

Anyway, I think you are right that libxl__device_pci_reset() should be retained in libxl__device_pci_add().  During startup, given "The real guest isn't using them yet, and the stubdom shouldn't use them on its own.", I think it's fine to just leave the reset and not try to re-work the reset flow.

Reply:
> Domain create calls the same functions as xl pci-attach. Domain create does:
Must have completely glossed over where those get called.  Interesting.

+1 for leaving flow as is for now.

Reply:
"Leave flow as is" meaning we keep the reset in libxl__device_pci_add(), yes?  That is what I am testing locally.

Reply:
Yea, to be clear:
- Leave the reset in pci add
- Also reset after the domain is destroyed